net/system_request_handler.h

```diff
@@ -0,0 +1,54 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_NET_SYSTEM_REQUEST_HANDLER_H_
+#define BRAVE_NET_SYSTEM_REQUEST_HANDLER_H_
+
+#include <memory>
+
+#include "base/callback.h"
+#include "net/base/net_export.h"
+
+namespace network {
+struct ResourceRequest;
+}  // network
+
+namespace base {
+template <typename T>
+struct DefaultSingletonTraits;
+}  // namespace base
+
+namespace net {
+
+class NET_EXPORT SystemRequestHandler {
```

Comment:
Tbh I don't see much sense in writing all this boilerplate for a single callback. IMO having one free function `OnBeforeSystemRequest` that calls something from a global variable should be enough

Reply:
This class can potentially be used in future for avoiding layering issues. @bridiver suggested to use singleton, but maybe I misunderstood it.
@iefremov do you insist on changing it?

Reply:
@iefremov this entire PR is to fix the layering violation and this class is how we do that

Reply:
we have to separate the callbacks themselves (which have `brave/browser` deps) from the method we call in `components/safe_browsing` which cannot have `brave/browser` deps

Reply:
@bridiver I'm saying that we could fix the layering violation *without* writing this class - since it adds some unnecessary boilerplate which takes some mental energy to understand. I'm not strongly against this singletone, but at least we should write comments explaining it's purpose.